File: python/luhn/luhn.py
__author__ = 'Cedric Zhuang'


class Luhn(object):
    def __init__(self, number):
        self.arr = map(int, str(number))

    @classmethod
    def create(cls, number):
        ret = number * 10
        luhn = Luhn(ret)
        to_add = (10 - luhn.checksum()) % 10
        return ret + to_add

    def addends(self):
        ret = self.arr[:]
        length = len(self.arr)
        for i in xrange(length-2, -1, -2):
            n = self.arr[i] * 2
            if n >= 10:
                n -= 9
            ret[i] = n
        return ret

    def checksum(self):
        return sum(self.addends()) % 10

    def is_valid(self):
        return self.checksum() == 0

File: scala/luhn/src/main/scala/Luhn.scala
/**
  *
  * @author Cedric Zhuang
  *
  */
case class Luhn(n: Long) {
  private def calc(c: Char, i: Int): Long = {
    val l: Long = c - '0'
    if (i % 2 == 0) l
    else if (l * 2 >= 10) l * 2 - 9
    else l * 2
  }

  lazy val checkDigit: Long = n % 10

  lazy val addends: List[Long] =
    n.toString.reverse.zipWithIndex
      .map { case (c, i) => calc(c, i) }
      .reverse.toList

  lazy val checksum: Long = addends.sum % 10

  lazy val isValid: Boolean = Luhn(n).checksum == 0

  lazy val create: Long = (10 - Luhn(n * 10).checksum) % 10 + n * 10
}
